### ISSUE TITLE
Surface Book 2 (i5) - Sudden heavy lag when gaming

### ISSUE THREAD

**Issue by SvenDaHacker64** [jakeday/linux-surface#458]
While I was playing games on [Citra - 3DS emulator](https://citra-emu.org/), it started to lag at random times where it goes from 60FPS down to 15FPS, then goes back to normal framerate, repeatedly. Happens on other games/emulators too. I'm using Ubuntu 18.04 / 19.04 on SB2 (only Intel graphics, no Nvidia).

It seems to be a similar behaviour on Windows 10 when you press the Detach button but didn't pull the tablet out. I didn't even hit the detach button on Ubuntu.

**Comment by qzed** [comment]
It looks like this is a heat issue (you could check that by looking at the dmesg log). From what I know, the i5 models are all 13.5", right? Problem is that the 13.5" models don't have a CPU fan, so there might not be much you can do. You could try setting a different performance-mode as described in https://github.com/qzed/linux-surfacegen5-acpi/issues/18, but I don't think that this will change much.

You could maybe also try to limit your cpu-clock using `cpufreq`. This would give you an overall lower clock-speed but might prevent the CPU from throttling down.

**Comment by SvenDaHacker64** [comment]
@qzed I'll give it a shot with cpufreq. Which one is recommended, the indicator-cpufreq or the GNOME extension?

**Comment by qzed** [comment]
I personally find the gnome extension quite useful, but if you want to use something on the command-line, you can use `cpupower frequency-set --max <...>`.

**Comment by SvenDaHacker64** [comment]
Cool, thanks. Hopefully it actually deals with throttling/slowdown.

**Comment by SvenDaHacker64** [comment]
I'm gonna try [konor's cpufreq](https://extensions.gnome.org/extension/1082/cpufreq/). By limiting the CPU clock, I move the Maximum slider, right?

**Comment by qzed** [comment]
Yes

**Comment by SvenDaHacker64** [comment]
I set min frequency as 50%, max frequency as 60%-75%, played a game for about 10mins and it doesn't slow down at all. I think that really helps. :)

**Comment by SvenDaHacker64** [comment]
@qzed I think cpufreq caused something to break after installing required system files - certain apps failed to load config related files in .local/share (broken permissions I guess?).

I'll switch to cpupower, what's the package name so I can do `apt-get install`? And what is the example value for --max , I'd like to set max frequency as 65%.

**Comment by qzed** [comment]
Hmm, that's a bit odd. I think on Ubuntu the package is `linux-tools-common`, but I'm not that sure about it. `--max` specifies the frequency, you can prefix that with `kHz`, e.g. `cpupower frequency-set --max 3500kHz` would set the maximum frequency to 3.5GHz. See http://manpages.ubuntu.com/manpages/bionic/man1/cpupower-frequency-set.1.html for details.

**Comment by SvenDaHacker64** [comment]
Found it. http://www.thinkwiki.org/wiki/How_to_use_cpufrequtils

I installed cpufrequtils with `sudo apt-get install cpufrequtils`, and setting my max frequency and governor with `sudo cpufreq-set -u 2.1Ghz -g performance`.

**Comment by qzed** [comment]
Awesome, that'll do the trick too. Just as a side-note: There are a couple of utilities that basically do the same thing. You can even directly set the frequency via sysfs.

**Comment by versipellis** [comment]
Hey @qzed, this is a tad old, but `cpufreq` tosses me an "irqbalance detected" notice. The [documentation](http://konkor.github.io/cpufreq/faq/) for `cpufreq` indicates it's a "safe to uninstall" for laptops (there's a discussion in the issue linked from the FAQ too about its advantages and disadvantages). I'm not sure if this is something that's critical for this kernel to run, though - did you uninstall it?

**Comment by qzed** [comment]
@versipellis I don't have `irqbalance` installed, so I think it's safe to uninstall. From what I can tell it's a tool to better schedule interrupts across multiple cores (and seems to be completely optional), so as you said it might not make that much sense on Laptops (as opposed to e.g. storage-server hardware where such interrupts are a lot more frequent).

**Comment by versipellis** [comment]
Gotcha, thanks. I'm curious how that ended up installed on my system.

**Comment by qzed** [comment]
@versipellis Seems to be installed by default on some distributions.

**Comment by versipellis** [comment]
Aaah, okay. Thanks! You're not running Ubuntu I take it, then?

**Comment by qzed** [comment]
@versipellis Nope, I'm on Arch :-) I was running Debian a while back which seems to have it installed by default. Didn't know that before you brought this up, so it's safe to say I haven't noticed any difference with regards to this.

**Comment by versipellis** [comment]
Awesome, thanks :)

**Comment by versipellis** [comment]
(I'm honestly curious to see if this resolves some of the CPU issues I'd been seeing with this kernel/linux on the SB2 with high-compute tasks like data analysis)

**Comment by qzed** [comment]
@versipellis With a bit of tweaking you should get some more stable clocks in the long term, Although you have to sacrifice short-term boosts for that. Could give you more performance overall. Also have you already looked at the performance-mode setting? At least if you've got the 15" model (i.e. the one with fans) that should make a difference.

**Comment by remingtonlang** [comment]
To chime in, I've noticed the same issue. I shutoff irqbalance but haven't had a chance to test that yet. Disabling turbo mitigates the issue to an extent in gaming though still exists, for analytics or ML tasks however that doesnt help much. I also stumbled on this and immediately thought of us: https://www.phoronix.com/forums/forum/linux-graphics-x-org-drivers/nvidia-linux/1103555-running-nvidia-on-gnome-s-x-org-session-may-get-a-lot-smoother
